topics/bloomfilters.md

```diff
@@ -0,0 +1,255 @@
+
+### Default bloom properties
+
+These are the default bloom properties along with the commands and configs which allow customizing.
+
+<table width="100%" border="1" style="border-collapse: collapse; border: 1px solid black" cellpadding="8">
+<tr>
+<th width="20%">Property</th>
+<th width="20%">Default Value</th>
+<th width="30%">Command Name</th>
+<th width="30%">Configuration name</th>
+</tr>
+<tr>
+<td>Capacity</td>
+<td>100</td>
+<td>BF.INSERT, BF.RESERVE</td>
+<td>BF.BLOOM-CAPACITY</td>
+</tr>
+<tr>
+<td>False Positive Rate</td>
+<td>0.01</td>
+<td>BF.INSERT, BF.RESERVE</td>
+<td>BF.BLOOM-FP-RATE</td>
+</tr>
+<tr>
+<td>Scaling / Non Scaling</td>
+<td>Scaling</td>
+<td>BF.INSERT, BF.RESERVE</td>
+<td>BF.BLOOM-EXPANSION</td>
+</tr>
+<tr>
+<td>Expansion Rate</td>
+<td>2</td>
+<td>BF.INSERT, BF.RESERVE</td>
+<td>BF.BLOOM-EXPANSION</td>
+</tr>
+<tr>
+<td>Tightening Ratio</td>
+<td>0.5</td>
+<td>BF.INSERT</td>
+<td>BF.BLOOM-TIGHTENING-RATIO</td>
+</tr>
+<tr>
+<td>Seed</td>
+<td>Random Seed</td>
+<td>BF.INSERT</td>
+<td>BF.BLOOM-USE-RANDOM-SEED</td>
+</tr>
+</table>
+
+
+Since bloom filters have a default expansion of 2, this means any default creation as a result of `BF.ADD`, `BF.MADD`, `BF.INSERT` will be a scalable bloom filter. Users can create a non scaling bloom filter using `BF.RESERVE <filter-name> <error-rate> <capacity> NONSCALING` or by specifying `NONSCALING` in `BF.INSERT`. Additionally, the other default properties of a bloom filter creation can be seen in the table above and BF.INFO command response below. These default properties can be configured through configs on the bloom module.
+
+Example of default bloom filter information:
+
+```
+127.0.0.1:6379> BF.ADD default_filter item
+1
+127.0.0.1:6379> BF.INFO default_filter
+ 1) Capacity
+ 2) (integer) 100
+ 3) Size
+ 4) (integer) 384
+ 5) Number of filters
+ 6) (integer) 1
+ 7) Number of items inserted
+ 8) (integer) 2
+ 9) Error rate
+10) "0.01"
+11) Expansion rate
+12) (integer) 2
+13) Tightening ratio
+14) "0.5"
+15) Max scaled capacity
+16) (integer) 26214300
+```
+
+## Performance
+
+The bloom commands which involve adding items or checking the existence of items have a time complexity of O(n * k) where n is the number of hash functions used by the bloom filter and k is the number of elements being inserted. This means that both BF.ADD and BF.EXISTS are both O(n) as they only operate on one item.
+
+Since performance relies on the number of hash functions, choosing the correct capacity and expansion rate can be important. In case of scalable bloom filters, with every scale out, we increase the number of checks (using hash functions of each sub filter) performed during any add / exists operation. For this reason, it is recommended that users choose a capacity after evaluating the use case / workload to avoid several scale outs and reduce the number of checks.
```

Comment:
```suggestion
In case of scalable bloom filters, with every scale out, we increase the number of checks (using hash functions of each sub filter) performed during any add / exists operation. For this reason, it is recommended that users choose a capacity and expansion rate after evaluating the use case / workload to avoid several scale outs and reduce the number of checks.
```